src/array.rs

```diff
@@ -1260,57 +1208,108 @@ impl<T: Element, D> PyArray<T, D> {
     /// # Example
     ///
     /// ```
-    /// # #[macro_use] extern crate ndarray;
-    /// use numpy::PyArray;
-    /// pyo3::Python::with_gil(|py| {
-    ///     let array = PyArray::from_exact_iter(py, 0..9);
-    ///     let array = array.reshape([3, 3]).unwrap();
-    ///     assert_eq!(array.readonly().as_array(), array![[0, 1, 2], [3, 4, 5], [6, 7, 8]]);
+    /// use numpy::{npyffi::NPY_ORDER, PyArray};
+    /// use pyo3::Python;
+    /// use ndarray::array;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let array =
+    ///         PyArray::from_iter(py, 0..9).reshape_with_order([3, 3], NPY_ORDER::NPY_FORTRANORDER).unwrap();
+    ///
+    ///     assert_eq!(array.readonly().as_array(), array![[0, 3, 6], [1, 4, 7], [2, 5, 8]]);
+    ///     assert!(array.is_fortran_contiguous());
+    ///
     ///     assert!(array.reshape([5]).is_err());
     /// });
     /// ```
-    #[inline(always)]
-    pub fn reshape<'py, ID, D2>(&'py self, dims: ID) -> PyResult<&'py PyArray<T, D2>>
-    where
-        ID: IntoDimension<Dim = D2>,
-        D2: Dimension,
-    {
-        self.reshape_with_order(dims, NPY_ORDER::NPY_ANYORDER)
-    }
-
-    /// Same as [reshape](method.reshape.html), but you can change the order of returned matrix.
-    pub fn reshape_with_order<'py, ID, D2>(
+    ///
+    /// [PyArray_Newshape]: https://numpy.org/doc/stable/reference/c-api/array.html#c.PyArray_Newshape
+    pub fn reshape_with_order<'py, ID: IntoDimension>(
         &'py self,
         dims: ID,
         order: NPY_ORDER,
-    ) -> PyResult<&'py PyArray<T, D2>>
-    where
-        ID: IntoDimension<Dim = D2>,
-        D2: Dimension,
-    {
-        let dims = dims.into_dimension();
-        let mut np_dims = dims.to_npy_dims();
+    ) -> PyResult<&'py PyArray<T, ID::Dim>> {
+        let mut dims = dims.into_dimension().to_npy_dims();
         let ptr = unsafe {
             PY_ARRAY_API.PyArray_Newshape(
                 self.py(),
                 self.as_array_ptr(),
-                &mut np_dims as *mut npyffi::PyArray_Dims,
+                &mut dims as *mut npyffi::PyArray_Dims,
                 order,
             )
         };
         if ptr.is_null() {
             Err(PyErr::fetch(self.py()))
         } else {
-            Ok(unsafe { PyArray::<T, D2>::from_owned_ptr(self.py(), ptr) })
+            Ok(unsafe { PyArray::<T, ID::Dim>::from_owned_ptr(self.py(), ptr) })
+        }
+    }
+
+    /// Special case of [`reshape_with_order`][Self::reshape_with_order] which keeps the memory order the same.
+    #[inline(always)]
+    pub fn reshape<'py, ID: IntoDimension>(
+        &'py self,
+        dims: ID,
+    ) -> PyResult<&'py PyArray<T, ID::Dim>> {
+        self.reshape_with_order(dims, NPY_ORDER::NPY_ANYORDER)
+    }
+
+    /// Extends or truncates the dimensions of an array.
+    ///
+    /// This method works only on [contiguous][`Self::is_contiguous`] arrays.
+    /// Missing elements will be initialized as if calling [`zeros`][Self::zeros].
```

Comment:
Out of curiosity what does this do for object? Fill with null references?

Reply:
From the doc comments of `PyArray::zeros`:

> For elements with DATA_TYPE == DataType::Object, this will fill the array with valid pointers to zero-valued Python integer objects.

So more specifically

```
>>> a = np.zeros(1, dtype=object)
>>> a[0]
0
>>> type(a[0])
<class 'int'>
```

(`PyArray::new` correspoding to `np.empty` will put zero pointers into object arrays (instead of uninitialized values for primitives like integers), so these arrays are always safe to drop, but not to use.)

((I noticed this still references `Element::DATA_TYPE` which does not actually exist any more. Will resolve this as part of the copy-editing done for #309.))

Reply:
(This behaviour is yet another reason by `PyArray<PyObject, D>` works, but `PyArray<Py<T>, D>` with arbitrary `T` is really problematic.)